lib/manager/components/LabelEditor.js

```diff
@@ -0,0 +1,224 @@
+  createLabel: Function,
+  label: Label,
+  onDone?: Function,
+  projectId: string,
+  updateLabel: Function,
+  user: ManagerUserState
+}
+
+type State = {
+  labelIsNew?: boolean,
+  newLabel: Label,
+  validation: {
+    description: boolean,
+    name: boolean
+  }
+}
+
+export default class LabelEditor extends React.Component<Props, State> {
+    state = {
+      validation: {
+        name: true,
+        description: true
+      },
+      newLabel: {...this.props.label},
+      labelIsNew: false
+    }
+
+    componentDidMount = () => {
+      // If we didn't get a label passed in, declare the label to be new
+      if (Object.keys(this.props.label).length === 0) {
+        this.setState({
+          labelIsNew: true,
+          // Default to a nice orange
+          // we can safely set a partial object here
+          // $FlowFixMe
+          newLabel: {
+            color: '#FF9200'
+          }
+        })
+      }
+    }
+
+    userIsAdmin = (user: ManagerUserState) => {
+      const {permissions} = user
+      if (!permissions) return false
+      else return permissions.isApplicationAdmin() || permissions.canAdministerAnOrganization()
+    }
+
+  _onFormChange = ({target}: {target: HTMLInputElement}, alwaysValid: boolean) => {
+    const {name, value, type, checked} = target
+
+    const universalValue: string | boolean = type === 'checkbox' ? checked : value
+    // extra check for flow
+    const valid =
+      alwaysValid ||
+      (typeof universalValue === 'string' && universalValue.length > 0)
+
+    this.setState(
+      update(this.state, {
+        newLabel: { $merge: {[name]: universalValue} },
+        validation: { [name]: { $set: universalValue && valid } }
+      })
+    )
+  }
+
+  // compares the dirty label from the old one to determine
+  // which keys changed
+  _getChanges = () => {
+    const newLabel = this.state.newLabel
+    const oldLabel = this.props.label
+
+    const changes: any = {}
+    Object.keys(newLabel).map(k => {
+      if (newLabel[k] !== oldLabel[k]) {
+        changes[k] = newLabel[k]
+      }
+    })
+    return changes
+  }
+  _settingsAreUnedited = () => Object.keys(this._getChanges()).length === 0
+
+  _formIsValid = () => {
+    const {validation} = this.state
+    return Object.keys(validation).every(k => validation[k])
+  }
+
+  _onSaveSettings = () => {
+    const {labelIsNew, newLabel: label} = this.state
+    const {createLabel, updateLabel} = this.props
+
+    if (labelIsNew) {
+      // Insert project ID
+      label.projectId = this.props.projectId
+      createLabel(label)
+    } else {
+      updateLabel(label)
+    }
+
+    // This passed method will close the editor
+    this.props.onDone && this.props.onDone()
+  }
+
+  render () {
+    const { newLabel, validation } = this.state
+    const { user } = this.props
+
+    return <div>
+      <Panel>
+        <ListGroup fill>
+          <ListGroupItem>
+            <FormGroup
+              data-test-id='label-name'
```

Comment:
It isn't necessary to always insert a data-test-id. These are mainly used with datatools's end-to-end tests when a specific selector is needed to reference certain elements. So since there isn't any additional e2e tests made for these labels, I wouldn't worry about adding these.

Reply:
Ah sounds good thank you!